Enrich Action View Template types with symbols from Action Dispatch Mime.

kaspth opened the pull request:
Needed for https://github.com/rails/rails/pull/23085

Main point is that `Types::Type.types` wouldn't pick any Mime symbols from Action Dispatch. Add `Type::SET.symbols` that maps directly to `Mime::SET.symbols`.

The other commits were simplifying the internals and sparing some `to_sym` calls thrown about.

kaspth committed:
Make ref return the internal symbol.

We delegate to_sym to the internal symbol, which we've already called to_sym on in
initialize, so we don't need to do that.

We also know to_sym will never return a falsy value, so we'll never hit to_s.

Just return the symbolized symbol.

kaspth committed:
Spare to_sym call in `==`.

The @symbol has already been converted to a symbol in initialize, so no need to call to_sym
when comparing it.

Ditch early return for a simple unless statement.

kaspth committed:
Replace delegate calls with standard method defs.

Spares a to_sym call by aliasing to_sym to ref.

Then the delegate felt meager for one method; ditch and define method ourselves.

kaspth committed:
Remove register abstraction.

The template types is a private abstraction to fill in basic blanks from Action Dispatch's
mime types. As such we can modify the data structure ourselves.

kaspth committed:
Replace class attribute with SET constant.

We'll be using this to map over to Action Dispatch's Mime::Set.

kaspth committed:
Don't bother looking up the types.

If they aren't symbols, then they aren't likely to be in the set anyway.

kaspth committed:
Enrich the SET constant to respond to symbols.

Match `Mime::SET.symbols`.

maclover7 commented:
:+1: thanks for this!

maclover7 commented on actionview/lib/action_view/template/types.rb:
```diff
@@ -28,30 +28,30 @@ def initialize(symbol)
       end
+
+      def symbols
+        type_klass::SET.symbols
```
This isn't going to work with the `TypeMap` implementation above -- we need to be returning an array, not a `Set`.
<img width="663" alt="screen shot 2016-01-17 at 10 31 38 am" src="https://cloud.githubusercontent.com/assets/3020626/12378224/89abd1e0-bd05-11e5-979f-c3b07a8295e9.png">

maclover7 replied:
`Mime::SET.symbols` returns an array:
<img width="684" alt="screen shot 2016-01-17 at 10 33 56 am" src="https://cloud.githubusercontent.com/assets/3020626/12378235/de70b984-bd05-11e5-87f2-e43358d96a86.png">

kaspth replied:
Think you're misunderstanding something. We aren't calling `symbols` on `@symbols`, we're calling it on `TypeMap.new`:

```irb
kaspers-macbook-pro:actionview kasperhansen$ irb
requiirb(main):001:0> require 'set'
=> true
irb(main):002:0> class TypeMap
irb(main):003:1>   attr_reader :symbols
irb(main):004:1> 
irb(main):005:1*   def initialize
irb(main):006:2>     @symbols = Set.new([ :html, :text, :js, :css, :xml, :json ])
irb(main):007:2>   end
irb(main):008:1> end
=> :initialize
irb(main):009:0> SET = TypeMap.new
=> #<TypeMap:0x007fc5f10bb940 @symbols=#<Set: {:html, :text, :js, :css, :xml, :json}>>
irb(main):010:0> SET.symbols
=> #<Set: {:html, :text, :js, :css, :xml, :json}>
```

kaspth replied:
Figured it would work since Sets and Arrays are both Enumerables, but:

```irb
irb(main):012:0> formats = [ :rss ]
=> [:rss]
irb(main):013:0> formats &= SET.symbols
TypeError: no implicit conversion of Set into Array
	from (irb):13:in `&'
	from (irb):13
	from /Users/kasperhansen/.rbenv/versions/2.2.3/bin/irb:11:in `<main>'
```

maclover7 replied:
Yep, that was just an example. We'll still need to convert the Set to an array though.

kaspth replied:
Just to clear something up here. Generally duck typing is preferred, so we don't have to maintain type consistency with `Mime::SET.symbols`. But I'm changing to an Array because of https://github.com/rails/rails/pull/23085/files#diff-d5d92e14e3693ca777dad370615cdaa9R70, which I've just demonstrated above will fail.

kaspth committed:
Store the symbols as an array.

A Set can't be implicitly converted into an Array:

```
irb(main):012:0> formats = [ :rss ]
=> [:rss]
irb(main):013:0> formats &= SET.symbols
TypeError: no implicit conversion of Set into Array
    from (irb):13:in `&'
    from (irb):13
    from /Users/kasperhansen/.rbenv/versions/2.2.3/bin/irb:11:in `<main>'
```

Besides `Mime::SET.symbols` returns an Array, so we're closer to that.
